invisible_cities/cities/detsim_simulate_electrons_test.py

```diff
@@ -0,0 +1,148 @@
+                    field_names = """xs ys zs energies
+                                     wi fano_factor
+                                     lifetime drift_velocity
+                                     transverse_diffusion longitudinal_diffusion""")
+
+    return nt(*itemgetter(*nt._fields)(locals()))
+
+
+def test_generate_ionization_electrons(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    n_electrons = generate_ionization_electrons(f.energies, f.wi, f.fano_factor)
+
+    # test same lengths and all positive
+    assert len(n_electrons) == len(f.energies)
+    assert np.all(n_electrons >= 0)
+
+    # test is integer
+    assert issubclass(n_electrons.dtype.type, (np.integer, int))
+
+
+def test_generate_ionization_electrons_null_energies(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    # test for null energies
+    n_electrons = generate_ionization_electrons(np.zeros(10), f.wi, f.fano_factor)
+    assert np.all(n_electrons == np.zeros(10))
+
+
+def test_drift_electrons(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    n_ie = np.mean(f.energies / f.wi)
+    n_electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
+    drifted_electrons = drift_electrons(f.zs, n_electrons, f.lifetime, f.drift_velocity)
+
+    # test all >= 0 and <= n_electrons
+    assert np.all(drifted_electrons >= 0)
+    assert np.all(drifted_electrons <= n_electrons)
+
+    # test is integer
+    assert issubclass(drifted_electrons.dtype.type, (np.integer, int))
+
+
+def test_drift_electrons_extreme_lifetimes(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    n_ie = np.mean(f.energies / f.wi)
+    n_electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
+
+    # test lifetime = 0
+    drifted_electrons = drift_electrons(f.zs, n_electrons, 0, f.drift_velocity)
+    assert np.all(drifted_electrons == 0)
+
+    # test lifetime = inf
+    drifted_electrons = drift_electrons(f.zs, n_electrons, np.inf, f.drift_velocity)
+    assert np.all(drifted_electrons == n_electrons)
+
+
+def test_diffuse_electrons(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    n_ie = np.mean(f.energies / f.wi)
+    n_electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
+
+    (dxs, dys, dzs) = diffuse_electrons(f.xs, f.ys, f.zs, n_electrons, f.transverse_diffusion, f.longitudinal_diffusion)
+
+    # test same lengths
+    assert len(dxs) == len(dys) == len(dzs) == np.sum(n_electrons)
+
+
+def test_diffuse_electrons_negative_z(MChits_and_detsim_params):
+
+    f = MChits_and_detsim_params
+
+    n_ie = np.mean(f.energies / f.wi)
+    n_electrons = np.clip(np.random.normal(n_ie, n_ie**0.5, size=len(f.zs)), 0, None).astype(int)
+
+    # test z<0 clip
+    zs = np.copy(f.zs)
+    sel = np.random.choice(np.arange(0, len(zs)), size=np.random.randint(1, len(zs)))
```

Comment:
> there is no good reason why the test would give a false positive for a specific set of indices unless the function is coded with malice.

Te-he, you'd be amazed at the bugs that property-based testing has found over the years, bugs that you would imagine are only possible through malice, until they actually appear in production code through innocent error. That's one of the big features of property-based testing: it has a habit of finding failing test cases that humans assume would never need testing because it would be absurd for them to fail given the cases that the humans have already checked. 

However, you do make a valid point: you have to find a reasonable balance between effort invested, clarity of code and benefits gained, rather than religiously testing absolutely everything to the max.

If you *are* going to introduce randomness into tests, please try to do it with hypothesis, rather than ad-hoc calls to random number generators. Once again, make some costs/benefits analysis rather than taking this as religious dogma.